File: 프로그래머스/lv3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>

using namespace std;

vector<bool> check;

void dfs(int now, const vector<vector<int>> computers){
    if(check[now]){
        return;
    }
    check[now] = true;
    for(int i = 0; i < computers[now].size(); i++){
        if(computers[now][i] == 1){
            dfs(i, computers);
        }
    }
}

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    
    check.resize(n);
    
    for(int i = 0; i < n; i++){
        if(!check[i]){
            answer++;
            dfs(i, computers);
        }
    }
    
    return answer;
}

File: 프로그래머스/lv3/43162. 네트워크/네트워크.java
class Solution {
    
    private static boolean[] visit;
    private static int[][] computer;
    
    public int solution(int n, int[][] computers) {
        computer = computers;
        visit = new boolean[n];
        int answer = 0;
        for(int i = 0; i < n; i++){
            if(!visit[i]){
                answer++;
                dfs(i);
            }
        }
        
        return answer;
    }
    private static void dfs(int now){
        if(visit[now]){
            return;
        }
        visit[now] = true;
        for(int i = 0; i < computer[now].length; i++){
            if(computer[now][i] == 1){
                dfs(i);
            }
        }
    }
}